UI for recently sold products

AnirudhBhat opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #8571 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR includes UI for displaying the recent products section in the product selection screen found in order creation.

To understand how this is calculated, please refer to this post: pdfdoF-2At-p2

If there are no recent products, the entire section will be hidden. Also, while searching for products, the recent section will be hidden and shown back again when the search query is cleared.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

#### Recent products section displayed
1. Make sure to log into the store that has many orders paid for with some products.
2. Navigate to the Order creation screen (Orders -> click on + button)
3. Click on the "Add products" button
4. Make sure you see the recent products section

#### Recent products section is hidden when there are no recent products
1. Make sure to log into the store that has just simple payment orders where no products are added while creating the order.
2. Navigate to the Order creation screen (Orders -> click on + button)
3. Click on the "Add products" button
4. Make sure you do not see the recent products section.

#### Recent products section is hidden when searching for products
1. Make sure to log into the store that has many orders paid for with some products.
2. Navigate to the Order creation screen (Orders -> click on + button)
3. Click on the "Add products" button
4. Make sure you see the recent products section
5. Search for any products
6. Make sure the recent products section gets hidden
7. Clear the search query
8. Make sure the recent products section gets shown again.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://user-images.githubusercontent.com/1331230/229059392-021a5e78-cf36-47ec-8d2b-76d255e279b2.mp4





- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

AnirudhBhat committed:
Add string resources for recent products section heading.

AnirudhBhat committed:
Add recently sold products compose UI component.

AnirudhBhat committed:
Add logic to hide the recent products section if the users are searching for products.

AnirudhBhat committed:
Add test to verify that the recent products section is hidden when the users are searching for products.

AnirudhBhat committed:
Add test to verify that the recent products section is shown again when the users clear the search query.

AnirudhBhat committed:
Add condition to add item inside lazyColumn only if popular or recently paid products list are not empty or null.

AnirudhBhat committed:
fix a bug where the divider was missing for popular section.

AnirudhBhat committed:
Merge branch 'issue/8570-ui-popular-products' into issue/8571-ui-recently-sold-products

AnirudhBhat committed:
Merge branch 'issue/8570-ui-popular-products' into issue/8571-ui-recently-sold-products

AnirudhBhat committed:
Fix failing tests

AnirudhBhat committed:
Merge branch 'issue/8570-ui-popular-products' into issue/8571-ui-recently-sold-products

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorScreen.kt:
```diff
@@ -194,7 +226,7 @@ private fun PopularProductsList(
         }
 
-        state.popularProducts.forEachIndexed { index, popularProduct ->
+        productsList.forEachIndexed { index, popularProduct ->
```
```suggestion
        productsList.forEachIndexed { index, product ->
```
💡 It can be both popular and recent.

AnirudhBhat replied:
Agree https://github.com/woocommerce/woocommerce-android/pull/8690/commits/0c026c77964ee7ead3e829a2b0cbdca0bb6c8c68

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorScreen.kt:
```diff
@@ -173,16 +173,48 @@ private fun EmptyProductList(
+
+@Composable
+private fun displayProductsSection(
+    type: ProductType,
```
💡 In general, we use `type` to get the header's text only. Why not pass the string instead of the type and remove `ProductType` enum to simplify the code?

AnirudhBhat replied:
`type` is used to decide which product list to populate on the UI as well.

```
val (productsList, heading) = when (type) {
        ProductType.POPULAR -> Pair(
            state.popularProducts, stringResource(id = string.product_selector_popular_products_heading)
        )
        ProductType.RECENT -> Pair(
            state.recentProducts, stringResource(id = string.product_selector_recent_products_heading)
        )
    }
```

I feel like using string here is kinda risky and prone to errors. WDYT?

samiuelson reviewed (commented):
Good work, @AnirudhBhat! I've added a few code improvement ideas for your consideration.

❓ Other than that, there is an issue that I think is worth fixing - the recently sold items section shows duplicated products. See the screenshot below. In order to reproduce it just create a few orders containing the same product.
<img src=https://user-images.githubusercontent.com/4527432/230097953-e5099c1c-e0d4-41ff-9664-5d9e4704c872.jpg width=400/>

💡 It would be reasonable to create a unit test for this edge case as well.

AnirudhBhat committed:
Use argument captors second value for recent products unit tests.

Since firstValue will be having popular section products list values.

AnirudhBhat committed:
Write a failing unit test to verify a bug.

Bug is that when multiple same products are purchased, the recent section displays them all instead of just 1 product.

AnirudhBhat committed:
Fix the bug, Make the test pass.

AnirudhBhat committed:
Fix a bug in the way we sorted popular products.

AnirudhBhat committed:
Remove duplicate products from recently sold list by using product id.

AnirudhBhat committed:
Rename popularProduct to product in productsList.forEachIndexed

AnirudhBhat commented:
>The recently sold items section shows duplicated products

Good catch! I've fixed this and it should work fine now. I've added the corresponding unit test as well.

samiuelson reviewed (approved):
Looks good! Good work 🎉
